docs/RFCS/20191202_full_cluster_backup_restore.md

```diff
@@ -0,0 +1,392 @@
+nodes remain available until the restore is complete.
+
+Non-cluster restore creates the tables for the user-data tables at the start of
+the restore. These tables are in an OFFLINE state - inaccessible to the
+user[1]. If there is a failure during the restore, and these tables are marked
+as DROP and will be removed. Full cluster restore can recover the user-data
+tables that it restored this way as well. The difficulty lies in handling
+system tables that it has already restored. This will likely be only the
+`system.zones` table since the remainder of the system tables will be restored
+in a single transaction near the end of the job, however the general case is
+considered.
+
+Since full cluster backup must have been run on a fresh cluster, the first
+iteration of full cluster restore could require the cluster to be destroyed if
+the restore fails. This can likely be improved as detailed in the Future Work
+section.
+
+
+# Alternatives Considered
+
+## System Table Restorations
+
+### AddSSTable and TimeBoundedDelete
+
+One reasonable question is why doesn't CockroachDB load the system data the
+same way as the user-data tables. One difficulty that this would present is
+that user-data restoration happens on new tables, but the system tables in the
+new cluster already have data in them. This method directly ingests the
+SSTables for the system tables spans, then issues a `TimeBoundedDelete`. This
+does not yet exist, but can be implemented by leveraging
+`engine.MVCCClearTimeRange`, similarly to the `batcheval.RevertRange` command.
+This leaves the possibility of having a potentially dirty state in the system
+tables. Additionally, the keys in the SSTs would need to have their timestamp
+updated to some time greater than the time of the start of the restore.
+
+The reason we take an approach of loading SSTs directly into the storage layer
+for user data is that we typically expect a large volume of data. Additionally,
+we can ensure that this data is not needed or accessed by the user while it
+is being loaded. Since the size of the system tables is expected to be much
+smaller than the size of the user-data tables, there are no advantages to
+this approach and it is more complex.
+
+## Cluster Info Metadata
+
+### Only Look at Backup Contents
+Additionally, instead of marking a particular backup as "full cluster", the
+system tables that it holds could be examined. This would allow for previous
+backups that included the system table information to be restored via a full
+cluster backup. One problem with this approach is that if new system tables are
+added to the list of expected system tables, a mapping between version numbers
+and which tables are expected to be included in that version needs to be
+maintained. This problem is avoided by marking backups as full cluster since we
+can assume that all system tables included in those backups are safe to restore
+(and override the existing ones).
+
+# Future Work
```

Comment:
Could you also mention multi-tenancy here?

https://github.com/cockroachdb/cockroach/issues/47970

I think this should mostly "fall into place", but worth checking if anything sticks out.